### PR TITLE
Glass Floors

### PR DESCRIPTION
**Pull request by N3X15**
# Executive Summary
Adds glass and plasmaglass flooring, which displays parallax space and piping (that hasn't been fucked up by mappers) below it.  Flooring is fragile and damaged if enough damage is dealt to it.  Heavy weights (fat people, mechs) damage flooring due to weight.

# Recipes
* lattice + reinforced glass = regular glass floor
* lattice + reinforced plasma glass = plasma glass floor.
* screwdriver -> crowbar -> welder => Lattice, (plasma) glass
# Gallery
Imgur
Imgur

# TODO
* [x] Glass shards aren't spawning due to a race condition.
* [x] Use old icon for turf as failover if parallax is disabled.
* [x] Deconstruction recipe.
* [x] Admin logging
* <s>[x] Falling</s> Animations are broken for various reasons, so temporarily abandoned.
* [x] Repair with silicate
* Additional heavy shit:
  * [x] Vehicles
  * [x] Mulebots

# Changelog
:cl:
* rscadd: Added glass floors.  Create by slapping steel lattice with reinforced glass or reinforced plasma glass. Fat people and mechs should keep off it.

### CODE REVIEW

**Comment by Megamaw** [comment]
>fat people are heavy enough to crack glass floors and space-age superglass floors

nexis with all due respect are you for serious right now

**Comment by N3X15** [comment]
@Megamaw yes

Because watching Rambling Maniac crash through a glass floor will never _not_ be funny.

**Comment by unid15** [comment]
Brilliant 👍👍👍

**Comment by Shadowmech88** [comment]
Will people ever see that? Intentionally breaching a room is bannable, so if a person is fat, and knows that fat people breach glass floors, why would the fat person ever walk over one and invite the banhammer?

**Comment by N3X15** [comment]
It would be most likely to happen after a bombing or during yakkety sax.

**Comment by clusterfack** [comment]
Why are the afterattack procs tabbed forward by a space?

Also you should make it compatible with non parallax space by giving it the space icon and setting the icon state.

**Comment by Exxion** [comment]
Oh yeah and the holodeck space sim should be replaced with a version of this with no glass overlay or breaking properties
Assuming it supports turfs other than holofloor at least

Though that would be a map change and thus be a separate PR I guess
Ignore me it's late

**Comment by strawberrytan** [comment]
You madman you actually did it thanks for being that guy

**Comment by strawberrytan** [comment]
And I don't know how to use GitHub properly but for the fat people breaking floors you could fatten up your victim and lock them in a glassfloored maintenance cubby

File: code/game/objects/items/stacks/sheets/glass.dm [hunk 1]
```diff
@@ -248,6 +248,36 @@
 	glass_quality = 1
 	shealth = 10
 
+	// Special snowflake code for building glass floors.
+	/obj/item/stack/sheet/glass/rglass/afterattack(atom/Target, mob/user, adjacent, params)
```
Comment (RemieRichards) [inline code/game/objects/items/stacks/sheets/glass.dm]: I like this new style of indentation.
almost artistic, or maybe autistic I'm not 100% sure.

**Comment by mtpowell** [comment]
Make the deconstruction the same as any other glass, screw > crowbar > screw > weld

**Comment by Lokiamis** [comment]
That's dope

File: code/game/objects/items/stacks/sheets/glass.dm (hunk 1 above)
Comment (N3X15) [reply code/game/objects/items/stacks/sheets/glass.dm]: both

**Comment by N3X15** [comment]
Can't unsecure turfs, so deconstruction recipe will have to be different.

**Comment by ghost** [comment]
What if falling down triggers the fat damage?
You can tiptoe through but watch out for banana peels, fattie.

**Comment by FudgePucker** [comment]
Sounds like good old fashioned family fun for everyone. What's the worst that could happen? :ok_hand:

**Comment by clusterfack** [comment]
You could make it so fat people can walk over it but not run over it

**Comment by Riekland** [comment]
There absolutely needs to be a normal varient
Like the sprite for it right now looks like what the plain glass would be and most certainly not the reinforced shit
Also rplasmaglass floors really shouldn't be breakable by fat people just so that you can use it in a public space without getting lynched

**Comment by N3X15** [comment]
The effect is ruined if you cover shit up with chickenwire or lattice.

**Comment by N3X15** [comment]
Dunno what the fuck is going on, but the animation doesn't seem to be taking.  Oh well.

**Comment by Blithering** [comment]
Personally, I'd prefer it if heavy things could only damage the glass if it's already damaged or loosened.
That way people can make corridors with a view without blocking all access to mules and mechs, but there's still possibility for sabotage funtimes and fat people shenanigans

**Comment by Durgen** [comment]
I think the falling through the z level thing is too inconsistent with the way space travel works in the rest of the game.
It's a bizzare mechanic that can and will be abused.
#keepspace2d

**Comment by ghost** [comment]
Hashtags dont work here

**Comment by Riekland** [comment]
Just add an edge thing like on carpets to the tile faget

**Comment by N3X15** [comment]
Maybe later.

**Comment by Aurx** [comment]
Currently, this doesn't appear to check for the absence of gravity.  Is it intended that fatsos will break the floor in 0g?

**Comment by N3X15** [comment]
N-no.

In my defense, however, when was the last time we had a zero-G event?

**Comment by Aurx** [comment]
I'm quite tempted to hit that button just so I could say "Now".  But yeah, it's basically never used.  

That said, I expect that eventually some jackass will try porting TGstation's gravity grief generator or something similar.

**Comment by nuklearcellphoneg** [comment]
Zero-G worst event.

Field generators for temporarily fixing floor breaches when

**Comment by Shadowmech88** [comment]
Don't the engineering vendors already come stocked with aluminum foam grenades that can be used for quickly patching floor breaches?

**Comment by N3X15** [comment]
Yeah, I'll fix it tomorrow.

**Comment by ultraplen** [comment]
How about reinforced glass floors ?

**Comment by N3X15** [comment]
Reinforced generally means "rods and shit criss-crossing the glass", which defeats the entire purpose of there being a glass fucking floor.

**Comment by Shadowmech88** [comment]
I can see objects beneath a full-tile reinforced window just fine.
I don't see why a reinforced glass floor should be terribly different.

**Comment by ghost** [comment]
sometimes some under-floor objects don't show. Just spawned some, and the pipes aren't showing either side of this manifold.

aspie spess junk mu 2017-07-02 160911

**Comment by ghost** [comment]
Current problems I've found

 - Some underfloor things such as pipes and wires don't show properly, as I mentioned above
 - Constructing the floor makes air from nowhere. Tested in space with walls around, make regular human pressure so would cause havoc in areas that require fine atmos control (Cold rooms, vox or plasmamen rooms, etc.). In comparison, making a plating tile spawns plating/airless so there is no magically appearing air.

`/obj/item/stack/tile/plasteel/proc/build(turf/S as turf)
	if(istype(S,/turf/space) || istype(S,/turf/unsimulated))
		S.ChangeTurf(/turf/simulated/floor/plating/airless)
	else
		S.ChangeTurf(/turf/simulated/floor/plating)`

 (Uncommented the animation stuff, spent a bit of time debugging it. I understand you've had enough of toying with it, so I'm generally just leaving this here for another PR)
 - The animation doesn't lock things to that turf for the duration, you can move around while 'falling'
 - It fails to actually 'throw that fucker' because of the changeturf, which means it tries to call /turf/space/proc/throw_that_fucker. Moving the contents of throw_that_fucker into where it is called after the spawn seems to work.
 - As it loops through all atom/movable on the turf, it throws things such as the virtualhearer, and any pipes that were on the turf. May be better to make a list of objects that were on the turf that actually exist.

**Comment by N3X15** [comment]
I think the pipes and shit are some sort of mapping issue.

**Comment by ghost** [comment]
They are there if you rightclick the turf and look at the contents. It's an odd thing

**Comment by N3X15** [comment]
Another reason I believe the missing pipes to be a mapping issue is that it only occurs with some pipes on the map, but those pipes are consistent, no matter the order of placement, nor whether the tiles were built, spawned, or placed in the map editor.

Anyway, deconstruction recipe is done and tested.

**Comment by Kammerjunk** [comment]
Can they be safely constructed in an atmosphere? It doesn't look like it.
If not, I'm not sure their weakness to grease golems is going to be what people immediately notice about them anyway - at least not until the guy working on them has finished fucking around with windows for their construction.

**Comment by ghost** [comment]
@Shadowmech88 There's only 1 or 2 and it's hacked iirc. Some maps have a box of them though, In addition though, basically everything will break the floor-only foam so for the most part there's no reason to use them over just making 2 wide plating and making it airtight with glass windows. You can make yourself or ask for super wazoo 3-LBluespace beaker 59 diameter ones for dudebombslmao but they're too rarely used and you give up making them/people start ignoring you asking for them pretty quick.

**Comment by Senaat1** [comment]
This is the worst fucking idea i've seen in a long time.

Merge that shit right now! 👍

**Comment by kilozombie** [comment]
Best lattice-related solution compromise I've found in the thread is that you can build glass floors without lattices to make them fatmannable, and build them on lattices (being a bit uglier) to be resistant to fat men.

**Comment by gbasood** [comment]
Someone's going to map glass floors into the bar
Then a wizard will show up and make everyone fat and no one will be able to leave
Alternatively someone will just get fat in the bar and step on the glass and murder everyone

**Comment by Shadowmech88** [comment]
@SarahJohnson489 
>basically everything will break the floor-only foam so for the most part there's no reason to use them

Do you think aluminum foam floors need to be made more durable?

**Comment by ghost** [comment]
@Shadowmech88 Yeah. Harm intent check like harm mech punch vs non-harm mech push but update in the desc. so it's not a code dive spoiler at the bare minimum. Maybe like 20 hp to avoid the first accidental click since you build over it you don't need to worry about the time to toolbox/extinguisher double tap it but antags can still toolbox new hole(s). It's a real downer when it's getting torn up left and right by people just clicking on this thing they've never seen before or by accident to the point that you wasted time bothering to throw it in the first place.

**Comment by ghost** [comment]
looks good to me

**Pull request merged**